design/proposals/modular-components.md

```diff
@@ -0,0 +1,163 @@
+# Modular Components and Cleaner Environments
+
+Status: Pending
+
+Version: Alpha
+
+## Abstract
+
+This document proposes to provide environments with an easier way of being
+selective of which components to interact with. It also proposes to make
+environments simpler.
+
+## Motivation
+
+The motivation behind this proposal is my user experience while trying to
+onboard Heptio's internal infrastructure configuration files onto ksonnet.
+
+The configuration files seem to commonly fall into one or more of these
+patterns:
+
+1. Configurations that are identical across environments (_i.e._, same in
+   _prod_, _dev_, and _staging_).
+2. Configurations that are nearly identical, however have simple parameters
+   that need to be tweaked across environments (_i.e._, _prod_ accessing a
+   different _mysql_ server than _dev_) 
+3. Configurations that are similar, however, have large pieces of code that
+   differ between environments and are not easily customizable by parameters.
+4. Configurations that are distinctively used in a single environment.
+
+ksonnet currently satisfies the use cases _1_ and _2_, however fall short on
+_3_ and _4_.
+
+## Goal
+
+* Redundancy between component manifests should be kept minimal.
+* It should be simple for an environment to only use a subset of all component
+  manifests.
+* It should be trivial to _add_ or _remove_ components from an environment.
+* It should be trivial to _clone_ an environment.
+* Environments should store the minimal amount of data necessary to operate
+  a set of _components_ on a _k8s cluster_.
+
+## Proposal
+
+1. Retire `main.jsonnet` and `base.libsonnet`.
+   The purpose of these files are often confusing to users. The goal of 
+   `main.jsonnet` was to act as a way for environments to provide
+   environment-specific extensibility to components. However, this only works
+   if the component is written in jsonnet, and also requires the user to have
+   an advanced understanding of jsonnet. Furthermore, with the existing model,
+   all components are operated on during `apply`, `delete`, etc. If the
+   components vary drastically between environments, it is not convenient to
+   "subtract" these components from each environment's `main.jsonnet`.
+   `base.libsonnet` also currently serves no purpose besides acting as an entry
```

Comment:
what do you think of removing `main.jsonnet` and `base.libsonnet` but still keeping env metadata within the env directories? 

e.g.  `env.yaml` (?) which includes the targets you mentioned (?) as well as server info

Reply:
because right now there's this strange split between `<env>/params.jsonnet` and the top-level `app.yaml`

Reply:
I don't have too strong of opinions for either approach. I think the benefits of keeping it in `app.yaml` is the ease to see configurations in one place (especially when the configuration isn't very long). And the cons are, this will keep the environments directory quite bare -- need to think more on what purpose the directory structure serves, and how to repurpose `params.libsonnet`, if needed.